.github/workflows/selfhosted.yml

```diff
@@ -0,0 +1,120 @@
+# CI workflow that runs on self hosted cluster
+name: Self-hosted rules
+
+on:
+  workflow_call:
+    inputs:
+      used-labels:
+        description: 'Label names in JSON format'
+        required: true
+        type: string
+      python-version:
+        description: 'Used python version'
+        required: true
+        type: string
+      slurm-job:
+        description: 'Name of the slurm job'
+        required: true
+        type: string
+      server-url:
+        description: ''
+        required: true
+        type: string
+      repository:
+        description: ''
+        required: true
+        type: string
+      run-id:
+        description: ''
+        required: true
+        type: string
+      poetry-extras:
+        description: 'poetry extra flags'
+        required: false
+        type: string
+        default: ""
+    secrets:
+      repo_token:
+        required: true
+
+jobs:
+  prepare:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - id: set-matrix
+        env:
+          LABELS: ${{ inputs.used-labels }}
+        run: |
+          platforms="{\"platform\":["
+          combined=""
+          shopt -s lastpipe
+          jq -c '.[]' <<< "$LABELS" | while read label; do
+              platform=(${label//-/ })
+              if [ ${platform[0]} == "\"run" ] && [ ${platform[1]} == "on" ]; then
+                platforms="${platforms}${combined:+,}\"${platform[2]}"
+                combined=${platforms}
+              fi
+          done
+          platforms+="]}"
+          echo ${platforms}
+          echo matrix="${platforms}" >> $GITHUB_OUTPUT
+  build:
+    needs: prepare
+    strategy:
+      matrix: ${{ fromJSON(needs.prepare.outputs.matrix) }}
+      fail-fast: false
+    runs-on: self-hosted
+    steps:
+    - name: Cleanup workspace manually
+      run: |
+        rm -rf _work/*
+    - uses: actions/checkout@v3
+    - name: Set up Python
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ inputs.python-version }}
+    - name: Install and configure poetry
+      uses: snok/install-poetry@v1
+      with:
+        virtualenvs-create: false
+    - name: Load cached venv
+      id: cached-pip-wheels
+      uses: actions/cache@v3
+      with:
+        path: ~/.cache
+        key: venv-${{ runner.os }}-${{ steps.setup-python.outputs.python-version }}-${{ hashFiles('**/poetry.lock') }}
+    - name: Install package
+      run: |
+        python -m venv testenv
+        source testenv/bin/activate
+        poetry install --no-interaction ${{ inputs.poetry-extras }}
```

Comment:
On one side, I'd avoid increasing too much the depth and complexity of the workflows.
But on the other, I've seen this snippet over and over, so I really wonder if we should split this.

I wonder if we can start defining actions to tame the complexity of workflows dependencies, but I fear they wouldn't be much simpler, and I wonder how to use an action from another action...
(i.e. most likely it is possible in JS actions, using some toolkit, but for this kind of things it would be simpler to define plain Bash actions)